### PR TITLE
Speed up MSVC Compilation

### PR DESCRIPTION
**Pull request by DavidFair**
Some fixes for MSVC I completed whilst working on another PR on the same platform:

- Compile using all cores available instead of 1, we should see compile times drop significantly on CI now
- Pass LTCG flag since the linker would start, realise we needed IPO then restart wasting ~5 seconds per target
- Bump CMake to 3.13 (this is for the above)
- Move IPO handling into its own CMake file, since the logic was slowly growing in the main CMake file
- ~Update CI to use Kitware APT repo as 3.10 is the latest available from the Ubuntu repo~ Azure uses 3.16 by default so we don't need to do this

Edit:

The following have been moved to their own issue and PR:

- Replace `Rebin` and the logical ops implicit conversions to static_cast (ensuring it's safe) reducing some warning noise. See commits for more details.
- Handle `bool` types specially in Rebin (since we were doing +=), this prevents an extremely unlikely bug where the int could wrap back up to 0 and turn into false and removes some warnings

Fixes: #1153

### CODE REVIEW

File: core/include/scipp/core/element/rebin.h [hunk 1]
```diff
@@ -13,6 +14,28 @@
...
+
+  using OldElementT = std::remove_reference_t<decltype(data_old[0])>;
+  using NewElementT = std::remove_reference_t<decltype(data_new[0])>;
+
+  if constexpr (std::is_same_v<bool, OldElementT>) {
+    // Bool -> T is well defined so no cast required
+    data_new[inew] = data_old[iold];
+  } else if constexpr (std::is_same_v<bool, NewElementT>) {
+    // use static_cast to ensure T -> Bool can be done at compile time
+    data_new[inew] = static_cast<NewElementT>(data_old[iold]);
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/rebin.h:31]: Why is this required? I don't recall supporting `rebin` with transistions between types?

Also, it looks like you are introducing some bugs here, you are now ignoring a lot of masking values and just default to use that of the last bin contributing to an output bin?

File: core/include/scipp/core/element/rebin.h [hunk 2]
```diff
@@ -37,13 +60,19 @@ static constexpr auto rebin = overloaded{
...
-                            std::decay_t<decltype(data_old)>>) {
+
+          constexpr bool isValAndVariance =
+              is_ValueAndVariance_v<std::decay_t<decltype(data_old)>>;
+
+          if constexpr (isValAndVariance) {
+            // Avoid implicit promotion and demotion when multiplying by scale
+            const auto scale =
+                std::remove_reference_t<decltype(data_new.value[0])>(delta /
+                                                                     owidth);
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/rebin.h:71]: This is a very verbose fix. Can't we just change how `delta` is created, e.g., using `std::min(...)` without explicit template arg, or with the desired size?

File: variable/logical.cpp [hunk 3]
```diff
@@ -14,28 +14,40 @@ using namespace scipp::core;
...
 namespace scipp::variable {
 
-static constexpr auto or_op_ = overloaded{
-    [](const auto &var_, const auto &other_) -> bool { return var_ | other_; },
-    dimensionless_unit_check_return};
+static constexpr auto or_op_ =
+    overloaded{[](const auto &var_, const auto &other_) -> bool {
+                 return static_cast<decltype(other_)>(var_) | other_;
+               },
+               dimensionless_unit_check_return};
```
Comment (SimonHeybrock) [inline variable/logical.cpp:21]: Why are these changes required? This should only ever be called with `bool`?

File: core/include/scipp/core/element/rebin.h [hunk 4]
```diff
@@ -37,13 +60,19 @@ static constexpr auto rebin = overloaded{
...
             data_new.value[inew] += data_old.value[iold] * scale;
             data_new.variance[inew] += data_old.variance[iold] * scale;
           } else {
-            data_new[inew] += data_old[iold] * scale;
+            rebinPrimativeType(data_new, data_old, inew, iold, delta, owidth);
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/rebin.h:75]:
```suggestion
            if constexpr(std::is_same_v<decltype(data_new[0]), bool)
              data_new[inew] |= data_old[iold];
            else
              data_new[inew] += data_old[iold] * scale;
```
?

File: core/include/scipp/core/element/rebin.h (hunk 1 above)
Comment (DavidFair) [reply core/include/scipp/core/element/rebin.h:31]: On MSVC we are getting warnings since Rebin is transitioning between types, so I'll add a static assert that old / new use the same element type and see what fails to compile

👍 - I'll look into fixing that

File: core/include/scipp/core/element/rebin.h (hunk 2 above)
Comment (DavidFair) [reply core/include/scipp/core/element/rebin.h:71]: Creating is with desired size rather than the implicit double would be better, as we then get no mismatched type warnings

However how would you recommend naming that type? I personally don't like the verbosity either

File: variable/logical.cpp (hunk 3 above)
Comment (DavidFair) [reply variable/logical.cpp:21]: MSVC warns about bool and int comparisons, so I'll add a static assert that the LHS/RHS are both bool types and see what falls over

File: core/include/scipp/core/element/rebin.h (hunk 1 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element/rebin.h:31]: > On MSVC we are getting warnings since Rebin is transitioning between types

We are not? The output data type should always be the input data type. If other cases are compiled there must be a bug in the type list.

File: variable/logical.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply variable/logical.cpp:21]: Hmm, is it because we are using bit-wise operators, and it casts the bools to ints? Otherwise I would not see why, both inputs should be bool?!

File: core/include/scipp/core/element/rebin.h (hunk 4 above)
Comment (DavidFair) [reply core/include/scipp/core/element/rebin.h:75]: 👍 - this was my original implementation, but then we had mixed types (bool, double) to still handle according to the compiler

I'll add a static_assert to guarantee they use the same type and use this form since I prefer this too

File: core/include/scipp/core/element/rebin.h (hunk 2 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element/rebin.h:71]: Let us discuss this after understanding why you had to add the helper function for the other branch.

File: core/include/scipp/core/element/rebin.h (hunk 4 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element/rebin.h:75]: This must be a bug in how it is used. There should *not* be mixed types showing up here. Don't add hacks here, we should try to find out where the actual problem lies.

File: core/include/scipp/core/element/rebin.h (hunk 1 above)
Comment (DavidFair) [reply core/include/scipp/core/element/rebin.h:31]: I'll have a look, it's either instantiating the templates with extra types we didn't ask for, or there is a bug which is causing some conversions to happen on MSVC / that GCC does not warn about

File: variable/logical.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply variable/logical.cpp:21]: https://stackoverflow.com/questions/44565126/how-to-enable-c-warnings-for-bitwise-operators-with-boolean-arguments?

Comment (DavidFair) [reply variable/logical.cpp:21]: This is similar to the problem in rebin, the compiler warning shows a generated lambda with bool / int for example
I suspect fixing one will fix multiple places where MSVC is warning about this

Comment (SimonHeybrock) [reply variable/logical.cpp:21]: and https://stackoverflow.com/questions/24542/using-bitwise-operators-for-booleans-in-c

Comment (DavidFair) [reply variable/logical.cpp:21]: So it's probably worth changing these bitwise operators to logical operators too, since in SCIPP were only concerned about True / False instead of bit manipulation

Comment (SimonHeybrock) [reply variable/logical.cpp:21]: Interestingly:
```python
import numpy as np

a = np.array([True, False])
~a # array([False,  True])
```
That is, numpy behavior does not match Python behavior, so the operators we provide appear to be consistent with numpy (we may just need to implement them differently)? As mentioned on Slack, I suggest to move this to a different issue.

**Comment by DavidFair** [comment]
I've marked the problems raised as resolved, since they will be completed in a separate PR and the commits have been dropped
